### ISSUE TITLE
Remove catalogs and introduce filtered indexes

### ISSUE THREAD

**Issue by eriohl** [Starcounter/Starcounter#2279]
Remove catalogs and introduce filtered indexes.

**Comment by eriohl** [comment]
Remaining:

1. Obsolete layout garbage collection (will be done as part of #1233).
2. Column type conversion (postponed until client adaption).

Also dependent on #2160.

**Comment by eriohl** [comment]
Obsolete layout garbage collection is working. However, there are some issues with it and recovery: If you load a database from an optimized log where obsolete layouts are dropped and then the keys are rused you can no longer load the database from the log. This doesn't really work of course, so some additional work in this will be required.

**Comment by eriohl** [comment]
Since we're now starting work on rearranging the write we will postpone work on layout garbage collection. If we need this before this can be finished we'll implement this as within the context of the current solution as well.

Added a note on discussing details of column type conversion to PNEXT adoption issue (#2605).

**Comment by eriohl** [comment]
Decision made that layout garbage collection will not be delivered until multi-process solution is done (#2611).

**Comment by eriohl** [comment]
Done.
